File: rotting-oranges/rotting-oranges.java
class Solution {
    public int orangesRotting(int[][] grid1) {
        return orangeGridFunction(0, grid1);
    }

    public int orangeGridFunction(int min,int[][] grid){
        int m = grid.length;
        int n = grid[0].length;
        int[][] original = new int[m][n];
        int oranges = 0;
        
        for(int i = 0;i < m;i++){
            for(int j = 0;j < n;j++) {
                original[i][j] = grid[i][j];
                if(original[i][j] == 1) {
                    oranges++;
                }
            }
        }

        if(oranges == 0) {
            return min;
        }

        for(int i = 0;i < m;i++){
            for(int j = 0;j < n;j++) {
                if(grid[i][j] == 2 && original[i][j] == 2) {
                    if(i > 0) {
                        if(grid[i-1][j] == 1){
                            grid[i-1][j] = 2;
                            oranges--;
                        }
                    }
                    if(i < m-1) {
                        if(grid[i+1][j] == 1){
                            grid[i+1][j] = 2;
                            oranges--;
                        }
                    }
                    if(j > 0) {
                        if(grid[i][j-1] == 1){
                            grid[i][j-1] = 2;
                            oranges--;
                        }
                    }
                    if(j < n-1) {
                        if(grid[i][j+1] == 1){
                            grid[i][j+1] = 2;
                            oranges--;
                        }
                    }
                }
            }
        }

        if(oranges == 0) {
            return min+1;
        }

        if(isEqual(original,grid) && oranges != 0) {
            return -1;
        }

        return orangeGridFunction(min+1,grid);
    }

    public boolean isEqual(int[][] org, int[][] modified) {
        int m = org.length;
        int n = org[0].length;
        for(int i = 0;i < m; i++) {
            for(int j = 0;j < n;j++) {
                if(org[i][j] != modified[i][j]){
                    return false;
                }
            }
        }

        return true;
    }
}

File: 1036-rotting-oranges/rotting-oranges.py
from collections import deque
import math
class Solution(object):
    directions = [[0,1],[0,-1],[1,0],[-1,0]]
    def orangesRotting(self, grid):
        rotten = [[float("inf")] * len(grid[0]) for h in range(len(grid))]
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == 2:
                    visited = [[False] * len(grid[0]) for k in range(len(grid))]
                    self.bfs(grid,visited,rotten,i,j)
        
        max_minutes = 0
        for i in range(len(grid)):
            for j in range(len(grid[0])):
                if grid[i][j] == 1 or grid[i][j] == 2:
                    if rotten[i][j] == float("inf"):
                        return -1
                    else:
                        max_minutes = max(max_minutes,rotten[i][j])

        return max_minutes
        
    def bfs(self, grid, visited, rotten, row, column):
        q = deque()
        q.appendleft([row,column])
        rotten[row][column] = 0
        parent, children, minutes = 1, 0, 1
        while len(q) > 0:
            r,c = q.pop()
            parent -= 1
            for direction in self.directions:
                nr,nc = r + direction[0], c + direction[1]
                if nr >= 0 and nr < len(grid) and nc >= 0 and nc < len(grid[0]) and not visited[nr][nc]:
                    if grid[nr][nc] == 1:
                        q.appendleft([nr,nc])
                        rotten[nr][nc] = min(rotten[nr][nc],minutes)
                        visited[nr][nc] = True
                        children += 1
            if parent == 0:
                parent = children
                children = 0
                minutes += 1
        return

        